File: JuliaGenie/src/controllers/TasksController.jl
module TasksController

using Genie.Requests
using Genie.Renderer.Json
using ..TasksService

function create()
  try
      user_id = parse(Int, params(:id))
      payload = jsonpayload()
      title = get(payload, "title", "")
      description = get(payload, "description", "")

      task = TasksService.create_task(user_id, title, description)
      json(task, status = 201)
  catch e
      if isa(e, ErrorException) && e.msg == "User not found"
          json(Dict("error" => "User not found"), status = 404)
      else
          json(Dict("error" => string(e)), status = 400)
      end
  end
end

function index()
  try
      user_id = parse(Int, params(:id))
      tasks = TasksService.list_tasks(user_id)
      json(tasks)
  catch e
      if isa(e, ErrorException) && e.msg == "User not found"
          json(Dict("error" => "User not found"), status = 404)
      else
          json(Dict("error" => string(e)), status = 400)
      end
  end
end

function show()
  id = parse(Int, params(:id))
  task = TasksService.get_task(id)
  if task === nothing
      json(Dict("error" => "Task not found"), status = 404)
  else
      json(task)
  end
end

function update()
  try
      id = parse(Int, params(:id))
      payload = jsonpayload()
      title = get(payload, "title", "")
      description = get(payload, "description", "")

      task = TasksService.update_task(id, title, description)
      if task === nothing
          json(Dict("error" => "Task not found"), status = 404)
      else
          json(task)
      end
  catch e
      json(Dict("error" => string(e)), status = 400)
  end
end

function mark_done()
  id = parse(Int, params(:id))
  task = TasksService.mark_done(id)
  if task === nothing
      json(Dict("error" => "Task not found"), status = 404)
  else
      json(task)
  end
end

function delete()
  id = parse(Int, params(:id))
  TasksService.delete_task(id)
  Genie.Renderer.Json.json(nothing, status = 204)
end

end

File: RubyInRails/app/controllers/tasks_controller.rb
class TasksController < ApplicationController
  def create
    task = TaskService.create(params[:user_id], task_params)
    if task
      render json: task, status: :created
    else
      render json: { error: "Invalid input" }, status: :bad_request
    end
  end

  def index
    tasks = TaskService.list_by_user(params[:user_id])
    render json: tasks
  end

  def show
    task = TaskService.get(params[:id])
    render json: task
  end

  def update
    task = TaskService.update(params[:id], task_params)
    render json: task
  end

  def done
    task = TaskService.mark_done(params[:id])
    render json: task
  end

  def destroy
    TaskService.delete(params[:id])
    head :no_content
  end

  private

  def task_params
    params.require(:task).permit(:title, :description, :done)
  rescue ActionController::ParameterMissing
     params.permit(:title, :description, :done)
  end
end
